File: packages/template-php/src/PageCache.php
<?php

declare(strict_types=1);

namespace Polyspec\Template;

use Closure;

/** Stores final HTML pages independently from compiled template artifacts. */
final class PageCache
{
    /** @var array<string, array{html: string, expires_at: float|null}> */
    private array $entries = [];

    public function __construct(private readonly ?Closure $clock = null)
    {
    }

    public function get(string $key): ?string
    {
        $entry = $this->entries[$key] ?? null;
        if ($entry === null) {
            return null;
        }
        if ($entry['expires_at'] !== null && $entry['expires_at'] <= $this->now()) {
            unset($this->entries[$key]);
            return null;
        }
        return $entry['html'];
    }

    public function set(string $key, string $html, ?float $ttl): void
    {
        if ($ttl !== null && (!is_finite($ttl) || $ttl < 0)) {
            throw new \InvalidArgumentException('page cache ttl must be null or non-negative');
        }
        $this->entries[$key] = [
            'html' => $html,
            'expires_at' => $ttl === null || $ttl === 0.0 ? null : $this->now() + $ttl,
        ];
    }

    /** Returns a cached page or renders and stores it on a miss. */
    public function getOrSet(string $key, ?float $ttl, Closure $render): string
    {
        $cached = $this->get($key);
        if ($cached !== null) {
            return $cached;
        }
        $html = $render();
        $this->set($key, $html, $ttl);
        return $html;
    }

    public function delete(string $key): void
    {
        unset($this->entries[$key]);
    }
    public function clear(): void
    {
        $this->entries = [];
    }

    private function now(): float
    {
        return $this->clock === null ? microtime(true) : (float) ($this->clock)();
    }
}

File: packages/template-rust/src/page_cache.rs
//! Final HTML page cache, separate from compiled template artifacts.
use std::collections::HashMap;

#[derive(Debug, Clone)]
struct Entry {
    html: String,
    expires_at: Option<f64>,
}

/// Stores final rendered HTML pages independently from template artifacts.
#[derive(Debug, Default)]
pub struct PageCache {
    entries: HashMap<String, Entry>,
}

impl PageCache {
    /// Creates an empty page cache.
    pub fn new() -> Self {
        Self::default()
    }

    /// Returns an unexpired page at the supplied Unix timestamp.
    pub fn get(&mut self, key: &str, now: f64) -> Option<String> {
        let expired = self
            .entries
            .get(key)
            .and_then(|entry| entry.expires_at)
            .is_some_and(|expires| expires <= now);
        if expired {
            self.entries.remove(key);
            return None;
        }
        self.entries.get(key).map(|entry| entry.html.clone())
    }

    /// Stores a page. `None` and zero mean no expiration.
    pub fn set(&mut self, key: impl Into<String>, html: impl Into<String>, ttl: Option<f64>, now: f64) -> Result<(), String> {
        if ttl.is_some_and(|value| !value.is_finite() || value < 0.0) {
            return Err("page cache ttl must be null or non-negative".to_string());
        }
        let expires_at = ttl.filter(|value| *value > 0.0).map(|value| now + value);
        self.entries.insert(
            key.into(),
            Entry {
                html: html.into(),
                expires_at,
            },
        );
        Ok(())
    }

    /// Returns a cached page or renders and stores it on a miss.
    pub fn get_or_set<F>(&mut self, key: &str, ttl: Option<f64>, now: f64, render: F) -> Result<String, String>
    where F: FnOnce() -> Result<String, String> {
        if let Some(html) = self.get(key, now) { return Ok(html); }
        let html = render()?;
        self.set(key, html.clone(), ttl, now)?;
        Ok(html)
    }

    /// Removes one page.
    pub fn delete(&mut self, key: &str) {
        self.entries.remove(key);
    }

    /// Removes all pages.
    pub fn clear(&mut self) {
        self.entries.clear();
    }
}
